apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerNumeric.cs

```diff
@@ -0,0 +1,482 @@
+﻿// Copyright 2020 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System;
+using System.Globalization;
+using System.Linq;
+using System.Numerics;
+using System.Text.RegularExpressions;
+
+namespace Google.Cloud.Spanner.V1
+{
+    // TODO: Implement IFormattable?
+
+    /// <summary>
+    /// Representation of the Spanner NUMERIC type, which has 38 digits of precision,
+    /// and a fixed scale of 9 decimal places to the right of the decimal point
+    /// </summary>
+    public struct SpannerNumeric : IEquatable<SpannerNumeric>, IComparable<SpannerNumeric>, IComparable
+    {
+        private static readonly BigInteger s_maxValue = BigInteger.Parse("99999999999999999999999999999999999999");
+        private static readonly BigInteger s_minValue = -s_maxValue;
+
+        private static readonly BigInteger[] s_powersOf10 = Enumerable.Range(0, 28).Select(p => BigInteger.Pow(10, p)).ToArray();
+
+        private static readonly BigInteger s_integerScalingFactor = new BigInteger(1_000_000_000L);
+        // TODO: Don't require a 0 before the decimal point.
+        // TODO: Replace with manual validation if we find this is a performance bottleneck (as it was with field name validation).
```

Comment:
This is something that happened in BigQuery, see #4976. It's OK to leave this TODO here, but reference the PR and change the comment to something like 

> (as it was with BigQuery field name valitions see #4976).

Reply:
All done.